session-03/src/index.ts

```diff
@@ -0,0 +1,141 @@
+import express from 'express';
+import type { Request, Response, NextFunction } from 'express';
+import fs from 'fs';
+import path from 'path';
+import mime from 'mime-types';
+import {
+  MAX_HEADER_LENGTH,
+  MAX_HEADER_COUNT,
+  MAX_LENGTH,
+  MAX_ID_LENGTH
+} from './constants';
+import {
+  getBlobPath,
+  getMetaPath,
+  validateQuota,
+  saveMetadata,
+  loadMetadata,
+  deleteBlob
+} from './blobStorage';
+
+const app = express();
+const port = 4000;
+
+function isValidId(id: string): boolean {
+  return /^[a-zA-Z0-9._-]{1,200}$/.test(id);
+}
+
+function extractAllowedHeaders(headers: Record<string, string | string[] | undefined>): Record<string, string> {
+  const result: Record<string, string> = {};
+  for (const [key, value] of Object.entries(headers)) {
+    const val = Array.isArray(value) ? value[0] : value ?? '';
+    const k = key.toLowerCase();
+    if ((k === 'content-type' || k.startsWith('x-rebase-')) &&
+        key.length <= MAX_HEADER_LENGTH &&
+        val.length <= MAX_HEADER_LENGTH) {
+      result[key] = val;
+    }
+  }
+  if (Object.keys(result).length > MAX_HEADER_COUNT) {
+    throw new Error('Too many headers');
+  }
+  return result;
+}
+
+app.post('/blobs/:id', (req: Request, res: Response, next: NextFunction) => {
+  (async () => {
+    const id = req.params.id;
+    if (!isValidId(id)) return res.status(400).send('Invalid ID');
+
+    const contentLength = req.headers['content-length'];
+    if (!contentLength) return res.status(411).send('Missing Content-Length');
+
+    const size = parseInt(contentLength, 10);
+    if (isNaN(size) || size > MAX_LENGTH) return res.status(413).send('Payload too large');
+
+    const headers = extractAllowedHeaders(req.headers as Record<string, string>);
+    let metaSize = 0;
+    for (const [k, v] of Object.entries(headers)) {
+      metaSize += k.length + v.length;
+    }
+    if (size + metaSize > MAX_LENGTH) return res.status(413).send('Total size exceeds limit');
+
+    const blobPath = getBlobPath(id);
+    const tempPath = blobPath + '.tmp';
+    await fs.promises.mkdir(path.dirname(blobPath), { recursive: true });
+
+    const writeStream = fs.createWriteStream(tempPath);
+    let written = 0;
+    let errored = false;
+
+    req.on('data', chunk => {
+      written += chunk.length;
+      if (written > MAX_LENGTH) {
+        errored = true;
+        req.destroy(new Error('Payload too large'));
+      }
+    });
+
+    req.pipe(writeStream);
+
+    writeStream.on('close', async () => {
+      if (errored) {
+        await fs.promises.rm(tempPath, { force: true });
+        return res.status(413).send('Stream too large');
+      }
+
+      try {
+        await validateQuota(written);
+        await fs.promises.rename(tempPath, blobPath);
+        await saveMetadata(id, {
+          id,
+          headers,
+          createdAt: new Date().toISOString(),
+          updatedAt: new Date().toISOString()
```

Comment:
consider introducing a local variable, say `now`, and use it twice. this will make sure that when storing for the first time, both fields have exactly the same value.